drivers/xbee/xbee.c

```diff
@@ -0,0 +1,700 @@
+        }
+        gpio_clear(sleep_pin);
+    }
+    /* if reset pin is connected, do a hardware reset */
+    if (reset_pin < GPIO_NUMOF) {
+        gpio_clear(reset_pin);
+        hwtimer_wait(HWTIMER_TICKS(RESET_DELAY));
+        gpio_set(reset_pin);
+    }
+     /* put the XBee device into command mode */
+    hwtimer_wait(HWTIMER_TICKS(ENTER_CMD_MODE_DELAY));
+    _at_cmd(dev, "+++");
+    hwtimer_wait(HWTIMER_TICKS(ENTER_CMD_MODE_DELAY));
+    /* disable non IEEE802.15.4 extensions */
+    _at_cmd(dev, "ATMM2\r");
+    /* put XBee module in "API mode without escaped characters" */
+    _at_cmd(dev, "ATAP1\r");
+    /* apply AT commands */
+    _at_cmd(dev, "ATAC\r");
+    /* exit command mode */
+    _at_cmd(dev, "ATCN\r");
+
+    /* set default short address */
+    _set_addr(dev, dev->addr_short, 2);
+    /* load long address (we can not set it, as it is read only) */
+    _get_addr_long(dev, dev->addr_long, 8);
+    /* set default channel */
+    tmp[1] = 0;
+    tmp[0] = XBEE_DEFAULT_CHANNEL;
+    _set_channel(dev, tmp, 2);
+    /* set default PAN ID */
+    tmp[1] = (uint8_t)(XBEE_DEFAULT_PANID >> 8);
+    tmp[0] = (uint8_t)(XBEE_DEFAULT_PANID & 0xff);
+    _set_panid(dev, tmp, 2);
+
+    DEBUG("xbee: Initialization successful\n");
+    return 0;
+}
+
+int _send(ng_netdev_t *netdev, ng_pktsnip_t *pkt)
+{
+    xbee_t *dev = (xbee_t *)netdev;
+    size_t size;
+    size_t pos;
+    ng_netif_hdr_t *hdr;
+    ng_pktsnip_t *payload;
+
+    /* check device descriptor and packet */
+    if (pkt == NULL) {
+        return -ENOMSG;
+    }
+    if (dev == NULL) {
+        ng_pktbuf_release(pkt);
+        return -ENODEV;
+    }
+
+    /* figure out the size of the payload to send */
+    size = ng_pkt_len(pkt->next);
+    if (size > XBEE_MAX_PAYLOAD_LENGTH) {
+        DEBUG("xbee: Error sending data, payload length exceeds limit\n");
+        ng_pktbuf_release(pkt);
+        return 0;
+    }
+    /* get netif header check address length */
+    hdr = (ng_netif_hdr_t *)pkt->data;
+    if (!(hdr->dst_l2addr_len == 2 || hdr->dst_l2addr_len == 8)) {
+        ng_pktbuf_release(pkt);
+        return -ENOMSG;
+    }
+
+    /* acquire TX lock */
+    mutex_lock(&(dev->tx_lock));
+    /* put together the API frame */
+    dev->tx_buf[0] = API_START_DELIMITER;
+    dev->tx_buf[4] = 0;         /* set to zero to disable response frame */
+    /* set size, API id and address field depending on dst address length  */
+    if (hdr->dst_l2addr_len == 2) {
+        dev->tx_buf[1] = (uint8_t)((size + 5) >> 8);
+        dev->tx_buf[2] = (uint8_t)(size + 5);
+        dev->tx_buf[3] = API_ID_TX_SHORT_ADDR;
+        memcpy(dev->tx_buf + 5, ng_netif_hdr_get_dst_addr(hdr), 2);
+        pos = 7;
+    } else {
+        dev->tx_buf[1] = (uint8_t)((size + 11) >> 8);
+        dev->tx_buf[2] = (uint8_t)(size + 11);
+        dev->tx_buf[3] = API_ID_TX_LONG_ADDR;
+        memcpy(dev->tx_buf + 11, ng_netif_hdr_get_dst_addr(hdr), 8);
+        pos = 13;
+    }
```

Comment:
Maybe that is documented for xbee somewhere, but how is the source address for the packet selected? For the `netdev` API I introduced `NETCONF_SRC_LEN` for this.